### PR TITLE
Remove more core dependency from ReferencePath

### PR DESCRIPTION
**Pull request by agrare**
TODO:
- [x] handle arrays

### CODE REVIEW

**Comment by kbrock** [comment]
I also want more_core_extensions out since this is the only place where we use it.

I was concerned about accessing strings, hashes, or arrays.
It seemed like json supports all 3. (🍅 please)
Does this new code handle all 3?


Glad you moved the `@path` code into the initializer. I am wondering why we need the `map` stuff. I think I just don't understand it.

File: spec/workflow/reference_path_spec.rb [hunk 1]
```diff
@@ -1,7 +1,78 @@
...
+      let(:input)   { {"hello" => "world"} }
+
+      it "returns the input" do
+        expect(subject.get(input)).to eq(input)
+      end
+    end
+
+    context "with an array dereference" do
+      let(:payload) { "$['store'][0]['book']" }
+      let(:input)   { {"store" => [{"book" => "ASL For Dummies"}]} }
```
Comment (kbrock) [inline spec/workflow/reference_path_spec.rb:26]: 0 seems like an easy case.
For some reason, I want to see 1 working

```suggestion
      let(:payload) { "$['store'][1]['book']" }
      let(:input)   { {"store" => [{},{"book" => "ASL For Dummies"}]} }
```

**Comment by kbrock** [comment]
can we test:

1. The top level object is an array. (we only currently deal with hashes, and I think array is valid. To punt, we can pass in `$` as the path.
2. Is there a way to append to an array? (per the spec)

File: spec/workflow/reference_path_spec.rb (hunk 1 above)
Comment (agrare) [reply spec/workflow/reference_path_spec.rb:26]: :+1: done (still works)

**Comment by agrare** [comment]
> The top level object is an array. 

Absolutely, done.

> Is there a way to append to an array? (per the spec)

That is a great question I don't know

**Comment by agrare** [comment]
> > Is there a way to append to an array? (per the spec)
>
> That is a great question I don't know

So it looks like `[-1:]` would mean the last element of an array, but the spec for reference paths specifically says that: `The operators "@", ",", ":", and "?" are not supported` so that wouldn't be valid.  I was able to easily set e.g. `$.[1].title` to create a second element in an array that only had one before though which seems like a valid case.

File: lib/floe/workflow/reference_path.rb [hunk 2]
```diff
@@ -4,37 +4,60 @@ module Floe
...
 
         raise Floe::InvalidWorkflowError, "Invalid Reference Path" if payload.match?(/@|,|:|\?/)
+        @path = JsonPath.new(payload)
+                        .path[1..]
+                        .map { |v| v.match(/\[(?<name>.+)\]/)["name"] }
+                        .map { |v| v[0] == "'" ? v.delete("'") : v.to_i }
+                        .compact
+      end
+
+      def get(context)
```
Comment (Fryguy) [inline lib/floe/workflow/reference_path.rb:29]: I would have expected get to just be .dig?

File: lib/floe/workflow/reference_path.rb [hunk 3]
```diff
@@ -4,37 +4,60 @@ module Floe
...
         result = context.dup
 
-        path = JsonPath.new(payload)
-                       .path[1..]
-                       .map { |v| v.match(/\[(?<name>.+)\]/)["name"] }
-                       .map { |v| v[0] == "'" ? v.delete("'") : v.to_i }
-                       .compact
-
         # If the payload is '$' then merge the value into the context
         # otherwise use store path to set the value to a sub-key
```
Comment (Fryguy) [inline lib/floe/workflow/reference_path.rb:45]: Minor but this comment refers to store path (as text so maybe it's ok)

**Comment by Fryguy** [comment]
High level is the goal just to drop more core?  If so I'm curious why?

**Comment by agrare** [comment]
Just trying to cut down on dependencies particularly because more_core brings in activesupport and trying to keep this "clean"

**Comment by kbrock** [comment]
If you ignore the development gems in `Gemfile`, dropping `more_core_extensions` drops quite a few gems:

- more_core_extensions
  - active_support
    - base64
    - bigdecimal
    - concurrent-ruby
    - connection_pool
    - drb
      - ruby2_keywords
    - i18n
    - minitest
    - mutex_m
    - tzinfo
  - sync

---

ASIDE: kubeclient has even more dependencies, but not trivial to remove:

- kubeclient
  - http
    - addressable
      - public_suffix
    - http-cookie
      - domain_name
    - http-form_data
    - llhttp-ffi
      - ffi-compiler
        - ffi
  - jsonpath
  - recursive-open-struct
  - rest-client
    - http-accept
    - (http-cookie)
    - mime-types
      - mime-types-data
    - netrc

**Comment by kbrock** [comment]
bump @Fryguy - you ok with this?

**Pull request merged**